startup/11-temperature-controller.py

```diff
@@ -256,10 +256,22 @@ class LinkamFurnace(PVPositioner):
     done = C(EpicsSignalRO, 'STATUS')
     stop_signal = C(EpicsSignal, 'RAMP:CTRL:SET')
     temperature = C(EpicsSignal, "TEMP")
+    
 
-    def set(self, *args, timeout=None, **kwargs):
+    def __init__(self, *args, setpoint_tolerance=1, **kwargs):
         
-        return super().set(*args, timeout=timeout, **kwargs)
+        # Sets tolerance for skipping `set` calls when already at setpoint.
+        self.setpoint_tolerance = setpoint_tolerance
+        super.__init__(*args, **kwargs)
+
+    def set(self, new_position, *args, timeout=None, **kwargs):
+
+        # Check if new setpoint is within one degree of current temperature, and if the controller is
+        # at it's setpoint. If it is, return that the move was successful instantly to avoid lockup.
+        if abs(self.readback.get() - new_position) < self.setpoint_tolerance and self.status.get() & 4 == 1:
```

Comment:
The way I designed this IOC is instead of making a PV for each the ~dozen or so status variables that are all just on/off switches, I have a general `STATUS` PV that is just a bitmask. The "At Setpoint" status value is the third bit in this bitmask, so if we do bitmask & 4 we will get a value of 1 if and only if the controller is at the setpoint.

Reply:
So that 1 doesn't need to be parametrized, it'll always be `1` for at setpoint and `0` for not at setpoint.

Reply:
Thanks for the explanation.